### PR TITLE
[SYCL][Graph] Enable L0 optimizations (no profiling mode)

### PR DESCRIPTION
**Pull request by mfrancepillois**
- Enable in-order cmd-list
Analyze the graph and apply enable the use of in-order command-list for linear graph. 
- Add a property to finalize function to enable graph profiling. 
- Update the specification.

### CODE REVIEW

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 1]
```diff
@@ -646,6 +651,15 @@ which is used in a graph will be kept alive on the host for the lifetime of the
...
 constructed with this property results in undefined behavior.
 
+===== Disable-In-Order-Optimization Property [[disable-in-order-optimization]]
+
+The `property::graph::disable_in_order_optimization` property disables 
+an optimization which can be applied to in-order graph to 
+reduce its execution time.
+This property is passed to the `command_graph::finalize()` function.
+Note that this property must be passed to the finalize function if users 
+want to profile the exec_graph execution.
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: This is quite a specific implementation detail to expose to the SYCL user, I think a better way to express the property would be `property::graph::enable_profiling` to align with queues, without going into the details of optimizations that will be disabled as there could be other optimizations in future that won't be compatible.

We should also specify an error is thrown when trying to profiling an event from a graph submission that was created without this property being set on the exec graph.

File: sycl/source/detail/graph_impl.cpp [hunk 2]
```diff
@@ -613,6 +614,8 @@ std::vector<sycl::detail::EventImplPtr> graph_impl::getExitNodesEvents() {
 void exec_graph_impl::findRealDeps(
     std::vector<sycl::detail::pi::PiExtSyncPoint> &Deps,
     std::shared_ptr<node_impl> CurrentNode, int ReferencePartitionNum) {
+  // if (MPartitions[MPartitionNodes[CurrentNode]]->MIsInOrderGraph)
+  //   return;
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.cpp]: can delete

File: sycl/source/detail/graph_impl.cpp [hunk 3]
```diff
@@ -694,6 +699,8 @@ void exec_graph_impl::createCommandBuffers(
 
   Partition->MPiCommandBuffers[Device] = OutCommandBuffer;
 
+  printGraphAsDot("execgraph.dot", false);
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.cpp]: debug code, can be removed

File: sycl/source/detail/graph_impl.hpp [hunk 4]
```diff
@@ -331,6 +331,24 @@ class node_impl {
     }
   }
 
+  /// Test is the node contains a N-D copy
+  /// @return true is the op is a N-D copy
+  bool is2DCopyNode() {
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.hpp]:
```suggestion
  bool is2DCopyNode() const {
```

File: sycl/source/detail/graph_impl.hpp [hunk 5]
```diff
@@ -550,13 +568,37 @@ class partition {
...
+  /// If so, the MIsInOrderGraph flag is set.
+  void checkIfGraphIsSinglePath() {
+    // Is the graph of this partition a single path graph?
+    // If so, we can optimize its execution using InOrder optimizations
+    MIsInOrderGraph = true;
+    if (MRoots.size() > 1) {
+      MIsInOrderGraph = false;
+      return;
+    }
+    for (auto Node : MSchedule) {
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.hpp]: Can avoid the copy here I think
```suggestion
    for (const auto &Node : MSchedule) {
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc (hunk 1 above)
Comment (mfrancepillois) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: I agree. A `property::graph::enable_profiling` could also cover other optimizations such as disabling profiling on events.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 6]
```diff
@@ -646,6 +650,17 @@ which is used in a graph will be kept alive on the host for the lifetime of the
 graph. Destroying that buffer during the lifetime of a `command_graph`
 constructed with this property results in undefined behavior.
 
+===== Enable-Profiling Property [[enable-profiling]]
+
+The `property::graph::enable_profiling` property enable profiling
+the executable_graph.
+This property is passed to the `command_graph::finalize()` function.
```
Comment (reble) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
The `property::graph::enable_profiling` property can be passed to the 
`command_graph::finalize()` function and enables profiling support 
for the returned `command_graph<graph_state::executable>`.
```

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 7]
```diff
@@ -849,8 +864,9 @@ Preconditions:
 
 Parameters:
 
-* `propList` - Optional parameter for passing properties. No finalization
-  properties are defined by this extension.
+* `propList` - Optional parameter for passing properties. The only defined 
+property is `disable_in_order_optimization`. See <<disable_in_order_optimization, Disable-In-Order-Optimization>>
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: old property name

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 8]
```diff
@@ -646,6 +650,17 @@ which is used in a graph will be kept alive on the host for the lifetime of the
...
+===== Enable-Profiling Property [[enable-profiling]]
+
+The `property::graph::enable_profiling` property enable profiling
+the executable_graph.
+This property is passed to the `command_graph::finalize()` function.
+If this property is not passed to the finalize function, certain optimizations,
+which prevent graph profiling, may be applied to this executable graph.
+In this case, an error will be thrown when attempting to profiling an event from 
+a graph submission that was created without this property.
+Note that this limitation currently only applies to the zero-level backend.
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: I think we should remove this line about level zero, and all backends should throw if this property isn't set for more consistent behaviour.

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp [hunk 9]
```diff
@@ -146,6 +146,14 @@ class assume_buffer_outlives_graph
 public:
   assume_buffer_outlives_graph() = default;
 };
+
+/// Property used to to add all previous graph leaves as dependencies when
+/// creating a new node with command_graph::add().
```
Comment (EwanC) [inline sycl/include/sycl/ext/oneapi/experimental/graph.hpp]: doxygen comment needs updated

File: sycl/source/detail/graph_impl.hpp [hunk 10]
```diff
@@ -1195,6 +1241,10 @@ class exec_graph_impl {
       MPartitionsExecutionEvents;
   /// Storage for copies of nodes from the original modifiable graph.
   std::vector<std::shared_ptr<node_impl>> MNodeStorage;
+  /// If true, the L0 backend In-order CommandList optimization is enabled.
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.hpp]: Is this optimization L0 specific, or will be applied to all backends?

File: sycl/test-e2e/Graph/event_profiling_info.cpp [hunk 11]
```diff
@@ -178,6 +182,25 @@ int main() {
     assert(verifyProfiling(CopyEvent) && verifyProfiling(KernelEvent1) &&
            verifyProfiling(KernelEvent2) &&
            compareProfiling(KernelEvent1, KernelEvent2));
+
+    // Checks exception thrown if profiling is requested while in-order
+    // optimization enabled. Note that in-order cmd-list optmization is only
+    // available for level-zero backend.
+    auto Backend = Dev.get_backend();
+    if (Backend == backend::ext_oneapi_level_zero) {
```
Comment (EwanC) [inline sycl/test-e2e/Graph/event_profiling_info.cpp]: See spec comment that I think all backends, not just L0, should throw if the enable_profiling property isn't set and profiling is done.

File: sycl/source/detail/graph_impl.hpp (hunk 10 above)
Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.hpp]: Yes, in-order optimization is level-zero specific. That said, I've removed this member because we do not really need it (MEnableProfiling is enough).

File: sycl/source/detail/graph_impl.hpp [hunk 12]
```diff
@@ -331,6 +331,24 @@ class node_impl {
     }
   }
 
+  /// Test is the node contains a N-D copy
```
Comment (julianmi) [inline sycl/source/detail/graph_impl.hpp]:
```suggestion
  /// Test if the node contains a N-D copy
```

File: sycl/source/detail/graph_impl.hpp [hunk 13]
```diff
@@ -331,6 +331,24 @@ class node_impl {
...
+  /// @return true is the op is a N-D copy
```
Comment (julianmi) [inline sycl/source/detail/graph_impl.hpp]:
```suggestion
  /// @return true if the op is a N-D copy
```

File: sycl/source/detail/graph_impl.hpp [hunk 14]
```diff
@@ -331,6 +331,24 @@ class node_impl {
...
+  bool is2DCopyNode() const {
```
Comment (julianmi) [inline sycl/source/detail/graph_impl.hpp]: Should we rename this to `isNDCopyNode` to reflect that we check for dimensions larger one?

File: sycl/source/detail/graph_impl.hpp [hunk 15]
```diff
@@ -550,13 +568,37 @@ class partition {
...
   bool isHostTask() const {
     return (MRoots.size() && ((*MRoots.begin()).lock()->MCGType ==
                               sycl::detail::CG::CGTYPE::CodeplayHostTask));
   }
 
+  /// Checks if the graph is single path, i.e. each node has a single successor.
+  /// If so, the MIsInOrderGraph flag is set.
+  void checkIfGraphIsSinglePath() {
+    // Is the graph of this partition a single path graph?
+    // If so, we can optimize its execution using InOrder optimizations
```
Comment (julianmi) [inline sycl/source/detail/graph_impl.hpp]: I think we can remove these comments since you cover this in the Doxygen comment already.

File: sycl/source/detail/graph_impl.hpp [hunk 16]
```diff
@@ -550,13 +568,37 @@ class partition {
...
 
   /// @return True if the partition contains a host task
   bool isHostTask() const {
     return (MRoots.size() && ((*MRoots.begin()).lock()->MCGType ==
                               sycl::detail::CG::CGTYPE::CodeplayHostTask));
   }
 
+  /// Checks if the graph is single path, i.e. each node has a single successor.
+  /// If so, the MIsInOrderGraph flag is set.
+  void checkIfGraphIsSinglePath() {
```
Comment (julianmi) [inline sycl/source/detail/graph_impl.hpp:590]: What is the overhead of adding this routine in comparison to the potential in-order optimization?

File: sycl/test-e2e/Graph/event_profiling_info.cpp [hunk 17]
```diff
@@ -130,8 +130,12 @@ int main() {
 
     KernelGraph.end_recording(Queue);
 
-    auto CopyGraphExec = CopyGraph.finalize();
-    auto KernelGraphExec = KernelGraph.finalize();
+    // The profiling is not available with the In-Order optimization.
```
Comment (julianmi) [inline sycl/test-e2e/Graph/event_profiling_info.cpp]:
```suggestion
    // The profiling is not available with the in-order optimization.
```

File: sycl/source/detail/graph_impl.hpp (hunk 16 above)
Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.hpp:590]: Its difficult to give an approximation of the overhead for this routine since it depends of the graph typology.
That said, if in-order graph is found we win on both sides: finalization delay (we do not need to create events) and execution time (we do not have to execute events nor synchronization on them).
On my setup (12th Gen Intel(R) Core(TM) i9-12900K, Intel(R) Level-Zero, Intel(R) UHD Graphics 770 1.3 [1.3.28454]), the finalization delay for an 2000 nodes in-order graph is reduced by ~40%. The execution time is reduced by ~15% and the second execution by ~20% compared to execution with event profiling capability disabled (and respectively 20% and 30% with the current implementation (i.e. event profiling enabled)).

Comment (julianmi) [reply sycl/source/detail/graph_impl.hpp:590]: Thanks, this sounds promising. I think we should run some microbenchmarks with and without these changes to better understand the overhead for nonlinear graphs.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 18]
```diff
@@ -341,6 +341,10 @@ class assume_buffer_outlives_graph {
   public:
     assume_buffer_outlives_graph() = default;
 };
+
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc:344]: missing `class enable_profiling {` with indent matching the above properties

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 19]
```diff
@@ -646,6 +650,18 @@ which is used in a graph will be kept alive on the host for the lifetime of the
...
+===== Enable-Profiling Property [[enable-profiling]]
+
+The `property::graph::enable_profiling` property can be passed to the 
+`command_graph::finalize()` function and enables profiling support 
+for the returned `command_graph<graph_state::executable>`.
+Passing this property to the finalize function implies disabling 
+certain optimizations to enable graph profiling.
+As a result, the execution time of a graph finalized with profiling enabled 
+is longer than that of a graph without profiling capability.
+In this case, an error will be thrown when attempting to profiling an event
```
Comment (EwanC) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]:
```suggestion
An error will be thrown when attempting to profile an event 
```

File: sycl/test-e2e/Graph/event_profiling_info.cpp [hunk 20]
```diff
@@ -178,6 +182,21 @@ int main() {
...
+        [&](handler &CGH) { CGH.ext_oneapi_graph(CopyGraphExecInOrder); });
+    Queue.wait_and_throw();
+    bool Success = false;
+    try {
+      EventInOrder
+          .get_profiling_info<sycl::info::event_profiling::command_start>();
+    } catch (sycl::exception &E) {
+      Success = true;
+    }
+    assert(Success);
```
Comment (EwanC) [inline sycl/test-e2e/Graph/event_profiling_info.cpp]: Think we normally put these kind of backend agnostic exception tests in `unittest`

File: sycl/source/detail/event_impl.hpp [hunk 21]
```diff
@@ -313,7 +315,7 @@ class event_impl {
   std::unique_ptr<HostProfilingInfo> MHostProfilingInfo;
   void *MCommand = nullptr;
   std::weak_ptr<queue_impl> MQueue;
-  const bool MIsProfilingEnabled = false;
+  bool MIsProfilingEnabled = false;
```
Comment (EwanC) [inline sycl/source/detail/event_impl.hpp:318]: Is this member ever set to `true`? I'm not getting any search hits for `MIsProfilingEnabled` or `setProfilingEnabled()` where anything other than `false` is used.

Comment (mfrancepillois) [reply sycl/source/detail/event_impl.hpp:318]: This member is set in the `event_impl` constructor (https://github.com/intel/llvm/blob/sycl/sycl/source/detail/event_impl.cpp#L159), but it must be reset if the graph profiling property is not set be users.

File: sycl/test-e2e/Graph/event_profiling_info.cpp (hunk 20 above)
Comment (mfrancepillois) [reply sycl/test-e2e/Graph/event_profiling_info.cpp]: I moved this test to `unittest`.

File: sycl/source/detail/graph_impl.hpp (hunk 16 above)
Comment (mfrancepillois) [reply sycl/source/detail/graph_impl.hpp:590]: On my setup, the `checkIfGraphIsSinglePath` function takes less than 0.01% of the total runtime of `finalize` for checking 2000 nodes.

File: sycl/source/detail/event_impl.hpp (hunk 21 above)
Comment (EwanC) [reply sycl/source/detail/event_impl.hpp:318]: makes sense ta, I misread the diff and thought we were adding this member

**Review by EwanC — APPROVED** [review]
LGTM 🙂

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp [hunk 22]
```diff
@@ -340,8 +349,10 @@ class __SYCL_EXPORT executable_command_graph {
   /// Constructor used by internal runtime.
   /// @param Graph Detail implementation class to construct with.
   /// @param Ctx Context to use for graph.
+  /// @param EnableProfiling Enable graph profiling.
   executable_command_graph(const std::shared_ptr<detail::graph_impl> &Graph,
-                           const sycl::context &Ctx);
+                           const sycl::context &Ctx,
+                           const bool EnableProfiling = false);
```
Comment (Bensuo) [inline sycl/include/sycl/ext/oneapi/experimental/graph.hpp]: In https://github.com/intel/llvm/pull/12840/ a similar change was made (adding a property on finalize) but the approach used was to forward the property list to this constructor. I have a preference for that approach (since it avoids the need to add new parameters every time for properties here) so unless you think this is better I would prefer if they were aligned.

File: sycl/source/detail/graph_impl.cpp [hunk 23]
```diff
@@ -680,7 +681,10 @@ sycl::detail::pi::PiExtSyncPoint exec_graph_impl::enqueueNode(
 void exec_graph_impl::createCommandBuffers(
     sycl::device Device, std::shared_ptr<partition> &Partition) {
   sycl::detail::pi::PiExtCommandBuffer OutCommandBuffer;
-  sycl::detail::pi::PiExtCommandBufferDesc Desc{};
+  sycl::detail::pi::PiExtCommandBufferDesc Desc{
+      pi_ext_structure_type::PI_EXT_STRUCTURE_TYPE_COMMAND_BUFFER_DESC, nullptr,
+      pi_bool(Partition->MIsInOrderGraph & !MEnableProfiling),
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.cpp]:
```suggestion
      pi_bool(Partition->MIsInOrderGraph && !MEnableProfiling),
```
Since we're dealing with `bool` here preference for `&&` since I don't believe `&` will perform short circuit evaluation.

File: sycl/source/detail/graph_impl.hpp [hunk 24]
```diff
@@ -550,13 +568,35 @@ class partition {
       MPiCommandBuffers;
   /// List of predecessors to this partition.
   std::vector<std::shared_ptr<partition>> MPredecessors;
+  /// True is the graph of this partition is a single path graph
+  /// and InOrder optmization can be applied on it.
```
Comment (Bensuo) [inline sycl/source/detail/graph_impl.hpp]:
```suggestion
  /// True if the graph of this partition is a single path graph
  /// and in-order optmization can be applied on it.
```

File: sycl/test-e2e/Graph/event_profiling_info.cpp [hunk 25]
```diff
@@ -130,8 +130,12 @@ int main() {
 
     KernelGraph.end_recording(Queue);
 
-    auto CopyGraphExec = CopyGraph.finalize();
-    auto KernelGraphExec = KernelGraph.finalize();
+    // The profiling is not available with the in-order optimization.
+    // We therefore disable this optimization.
```
Comment (Bensuo) [inline sycl/test-e2e/Graph/event_profiling_info.cpp]: This comment is a bit confusing, why are we talking about the in-order optimzation at all here? It seems irrelevant to this test.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc [hunk 26]
```diff
@@ -646,6 +651,18 @@ which is used in a graph will be kept alive on the host for the lifetime of the
...
+
+The `property::graph::enable_profiling` property can be passed to the 
+`command_graph::finalize()` function and enables profiling support 
+for the returned `command_graph<graph_state::executable>`.
+Passing this property to the finalize function implies disabling 
+certain optimizations to enable graph profiling.
+As a result, the execution time of a graph finalized with profiling enabled 
+is longer than that of a graph without profiling capability.
+An error will be thrown when attempting to profile an event 
+from a graph submission that was created without this property.
```
Comment (Bensuo) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: For Record & Replay, do we need to pass both the queue profiling property and the graph property to be able to profile graph events?

Comment (mfrancepillois) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: Yes, we need both in the current implementation.

File: sycl/test-e2e/Graph/event_profiling_info.cpp (hunk 25 above)
Comment (mfrancepillois) [reply sycl/test-e2e/Graph/event_profiling_info.cpp]: You right. I forgot to update this comment. Now I have.

File: sycl/include/sycl/ext/oneapi/experimental/graph.hpp (hunk 22 above)
Comment (mfrancepillois) [reply sycl/include/sycl/ext/oneapi/experimental/graph.hpp]: The code has been updated this way.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc (hunk 26 above)
Comment (Bensuo) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: Hmm, I'm not super keen on requiring both for R&R, some thoughts:

- Perhaps the property would be better served being for nodes in the explicit API? That way there is a clear single mechanism for each API, while also allowing for more granularity (not forcing profiling on all nodes).
- I'm guessing both are required because otherwise the returned events from command recording would have profiling disabled, but after finalize they would then have profiling enabled and this is confusing to the user? However, I don't like that passing the queue property but not the graph one is basically a no-op, I think that is more confusing.

File: sycl/source/detail/graph_impl.hpp (hunk 16 above)
Comment (reble) [reply sycl/source/detail/graph_impl.hpp:590]: Im not sure if just looking at the single function call is a fair metric, because the `Schedule` of the Graph is already available. That obviously is an implementation detail and won't necessarily take away the concerns about the complexity of this check. Another aspect that might be relevant: What if the check fails, it might be still beneficial to execute `Schedule` as-is on an in-order CommandList, or interleaving on multiple in-order CommandLists. That brings up the question if a scheduling hint that we'd pass as a property on graph finalization is a better option?

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc (hunk 26 above)
Comment (mfrancepillois) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: I've updated the code (and the spec) accordingly.
Now, users must enable the profiling on at least one node to enable the profiling of the graph,
This can be done by using either the `property::graph::enable_profiling` property in building mode or the `property::queue::enable_profiling` on the recorded queue (Record&Replay API).
Note that the `property::queue::enable_profiling` must also set on the queue used to submit the graph. The event submission timestamp is indeed managed by the runtime, and is only set when submit queue has the flag. This might change with the profiling tag...

Comment (mfrancepillois) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc]: It's not implemented yet, but I think it might be useful to add a new API to disable profiling on nodes. Such as `node::update_properties(const property_list &PropList)` with a new property `property::queue::disable_profiling`?

**Review by Bensuo — APPROVED** [review]

File: sycl/source/detail/graph_impl.hpp (hunk 16 above)
Comment (Bensuo) [reply sycl/source/detail/graph_impl.hpp:590]: Extending running things in-order to be user controlled with a property or something definitely seems like it could be useful. 

But as to the other point of discussion here, do we really need to care that much about small optimizations of `finalize()`? It is expensive by design and barring any outlandishly slow performance it seems to me it doesn't really matter much at all how it performs.

Comment (Bensuo) [reply sycl/source/detail/graph_impl.hpp:590]: Regarding the idea of adding a hint/property to enable in-order command lists in more situation, it seems that probably requires some more in-depth discussion and is probably better done as a separate PR to avoid delaying this one too much.

**Review by julianmi — APPROVED** [review]

**Comment by Bensuo** [comment]
Upstream PR here (draft until UR changes merge): https://github.com/intel/llvm/pull/13088
